pyomo/contrib/pynumero/sparse/mpi_block_vector.py

```diff
@@ -0,0 +1,1291 @@
+
+        Returns
+        -------
+        BlockVector
+        """
+        """
+        We do this by serializing the structure, then gathering it.
+        To serialize the structure, we use an array. The first number indicates if the first block is a numpy array 
+        or a BlockVector. We use -1 to indicate a BlockVector and -2 to indicate a numpy array. If the block is a
+        BlockVector, then the next number is a positive integer specifying the number of blocks in the block vector.
+        If the block is a numpy array, then the next number is a positive integer specifying the size of the array. 
+        After the number of blocks in a BlockVector is specified, we follow the same procedure to specify the 
+        structure of that BlockVector. 
+        """
+        assert_block_structure(self)
+        serialized_structure_by_block = dict()
+        length_per_block = np.zeros(self.nblocks, dtype=np.int64)
+        rank = self._mpiw.Get_rank()
+        if rank == 0:
+            block_indices = self._owned_blocks
+        else:
+            block_indices = self._unique_owned_blocks
+        for ndx in block_indices:
+            blk = self.get_block(ndx)
+            blk_structure = list()
+            if isinstance(blk, BlockVector):
+                blk_structure.append(-1)
+                blk_structure.append(blk.nblocks)
+                blk_structure.extend(self._serialize_structure(blk))
+            elif isinstance(blk, MPIBlockVector):
+                raise NotImplementedError('Operation not supported for MPIBlockVectors containing other MPIBlockVectors')
+            elif isinstance(blk, np.ndarray):
+                blk_structure.append(-2)
+                blk_structure.append(blk.size)
+            else:
+                raise NotImplementedError('Unrecognized input.')
+            length_per_block[ndx] = len(blk_structure)
+            serialized_structure_by_block[ndx] = np.asarray(blk_structure, dtype=np.int64)
+
+        global_length_per_block = np.zeros(self.nblocks, dtype=np.int64)
+        self._mpiw.Allreduce(length_per_block, global_length_per_block)
+        local_serialized_structure = np.zeros(global_length_per_block.sum(), dtype=np.int64)
+
+        offset = 0
+        block_indices_set = set(block_indices)
+        for ndx in range(self.nblocks):
+            if ndx in block_indices_set:
+                local_serialized_structure[offset: offset+global_length_per_block[ndx]] = serialized_structure_by_block[ndx]
+            offset += global_length_per_block[ndx]
+        global_serialized_structure = np.zeros(global_length_per_block.sum(), dtype=np.int64)
+        self._mpiw.Allreduce(local_serialized_structure, global_serialized_structure)
+
+        result = BlockVector(self.nblocks)
+        structure_ndx = 0
+        self._create_from_serialized_structure(global_serialized_structure, structure_ndx, result)
+
+        return result
+
+    def make_local_copy(self):
```

Comment:
same question as above. I think I made a comment about wanting this on MPIBlockMatrix.

Reply:
I created an issue for this.